pkg/api/service/annotations.go

```diff
@@ -25,4 +32,58 @@ const (
 	//
 	// Not all cloud providers support this annotation, though AWS & GCE do.
 	AnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
+
+	// AnnotationExternalTraffic An annotation that denotes if this Service desires to route external traffic to local
+	// endpoints only. This preserves Source IP and avoids a second hop.
+	AnnotationExternalTraffic = "service.alpha.kubernetes.io/external-traffic"
+	// AnnotationValueExternalTrafficLocal Value of annotation to specify local endpoints behaviour
+	AnnotationValueExternalTrafficLocal = "OnlyLocal"
+	// AnnotationValueExternalTrafficGlobal Value of annotation to specify global (legacy) behaviour
+	AnnotationValueExternalTrafficGlobal = "Global"
+	// AnnotationHealthCheckNodePort Annotation specifying the healthcheck nodePort for the service
+	// If not specified, annotation is created by the service api backend with the allocated nodePort
+	// Will use user-specified nodePort value if specified by the client
+	AnnotationHealthCheckNodePort = "service.alpha.kubernetes.io/healthcheck-nodeport"
```

Comment:
Yeah as long as this is feature gated.

On Tue, Sep 6, 2016 at 12:03 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/api/service/annotations.go
> <https://github.com/kubernetes/kubernetes/pull/29409#discussion_r77570919>
> :
>
> > @@ -25,4 +32,58 @@ const (
> >  	//
> >  	// Not all cloud providers support this annotation, though AWS & GCE do.
> >  	AnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
> > +
> > +	// AnnotationExternalTraffic An annotation that denotes if this Service desires to route external traffic to local
> > +	// endpoints only. This preserves Source IP and avoids a second hop.
> > +	AnnotationExternalTraffic = "service.alpha.kubernetes.io/external-traffic"
> > +	// AnnotationValueExternalTrafficLocal Value of annotation to specify local endpoints behaviour
> > +	AnnotationValueExternalTrafficLocal = "OnlyLocal"
> > +	// AnnotationValueExternalTrafficGlobal Value of annotation to specify global (legacy) behaviour
> > +	AnnotationValueExternalTrafficGlobal = "Global"
> > +	// AnnotationHealthCheckNodePort Annotation specifying the healthcheck nodePort for the service
> > +	// If not specified, annotation is created by the service api backend with the allocated nodePort
> > +	// Will use user-specified nodePort value if specified by the client
> > +	AnnotationHealthCheckNodePort = "service.alpha.kubernetes.io/healthcheck-nodeport"
>
> @girishkalele <https://github.com/girishkalele> - not urgent for Alpha,
> should handle in beta. Open a tracking issue and link it to the feature
> bug? On Mon, Sep 5, 2016 at 7:29 PM, Clayton Coleman <
> notifications@github.com> wrote:
> … <#m_5481768344954675212_>
> In pkg/api/service/annotations.go <#29409 (comment)
> <https://github.com/kubernetes/kubernetes/pull/29409#discussion_r77565957>>
> : > @@ -25,4 +32,58 @@ const ( > // > // Not all cloud providers support
> this annotation, though AWS & GCE do. > AnnotationLoadBalancerSourceRangesKey
> = "service.beta.kubernetes.io/load-balancer-source-ranges" > + > + //
> AnnotationExternalTraffic An annotation that denotes if this Service
> desires to route external traffic to local > + // endpoints only. This
> preserves Source IP and avoids a second hop. > + AnnotationExternalTraffic
> = "service.alpha.kubernetes.io/external-traffic" > + //
> AnnotationValueExternalTrafficLocal Value of annotation to specify local
> endpoints behaviour > + AnnotationValueExternalTrafficLocal = "OnlyLocal"
> > + // AnnotationValueExternalTrafficGlobal Value of annotation to
> specify global (legacy) behaviour > + AnnotationValueExternalTrafficGlobal
> = "Global" > + // AnnotationHealthCheckNodePort Annotation specifying the
> healthcheck nodePort for the service > + // If not specified, annotation is
> created by the service api backend with the allocated nodePort > + // Will
> use user-specified nodePort value if specified by the client > +
> AnnotationHealthCheckNodePort = "service.alpha.kubernetes.io/
> healthcheck-nodeport" k8s.io/kubernetes/pkg/quota/
> evaluator/core/services.go should contain the calculator. …
> <#m_2278866094229326858_> On Mon, Sep 5, 2016 at 7:24 PM, Tim Hockin
> ***@***.***> wrote: In pkg/api/service/annotations.go <#29409
> <https://github.com/kubernetes/kubernetes/pull/29409> (comment) <#29409
> (comment)
> <https://github.com/kubernetes/kubernetes/pull/29409#discussion_r77559231>>>
> : > @@ -25,4 +32,58 @@ const ( > // > // Not all cloud providers support
> this annotation, though AWS & GCE do. > AnnotationLoadBalancerSourceRangesKey
> = "service.beta.kubernetes.io/load-balancer-source-ranges" > + > + //
> AnnotationExternalTraffic An annotation that denotes if this Service
> desires to route external traffic to local > + // endpoints only. This
> preserves Source IP and avoids a second hop. > + AnnotationExternalTraffic
> = "service.alpha.kubernetes.io/external-traffic" > + //
> AnnotationValueExternalTrafficLocal Value of annotation to specify local
> endpoints behaviour > + AnnotationValueExternalTrafficLocal = "OnlyLocal"
> > + // AnnotationValueExternalTrafficGlobal Value of annotation to
> specify global (legacy) behaviour > + AnnotationValueExternalTrafficGlobal
> = "Global" > + // AnnotationHealthCheckNodePort Annotation specifying the
> healthcheck nodePort for the service > + // If not specified, annotation is
> created by the service api backend with the allocated nodePort > + // Will
> use user-specified nodePort value if specified by the client > +
> AnnotationHealthCheckNodePort = "service.alpha.kubernetes.io/
> healthcheck-nodeport" This allocates a max f one NodePort per service, but
> yeah. Any experts can point in the right direction? …
> <#m_-257820063610826413_> On Sep 5, 2016 11:58 AM, "Clayton Coleman"
> ***@***.***> wrote: In pkg/api/service/annotations.go <#29409
> <https://github.com/kubernetes/kubernetes/pull/29409> <#29409
> <https://github.com/kubernetes/kubernetes/pull/29409>> (comment) <#29409
> <https://github.com/kubernetes/kubernetes/pull/29409> (comment) <#29409
> (comment)
> <https://github.com/kubernetes/kubernetes/pull/29409#discussion_r77550810>>>>
> : > @@ -25,4 +32,58 @@ const ( > // > // Not all cloud providers support
> this annotation, though AWS & GCE do. > AnnotationLoadBalancerSourceRangesKey
> = "service.beta.kubernetes.io/load-balancer-source-ranges" > + > + //
> AnnotationExternalTraffic An annotation that denotes if this Service
> desires to route external traffic to local > + // endpoints only. This
> preserves Source IP and avoids a second hop. > + AnnotationExternalTraffic
> = "service.alpha.kubernetes.io/external-traffic" > + //
> AnnotationValueExternalTrafficLocal Value of annotation to specify local
> endpoints behaviour > + AnnotationValueExternalTrafficLocal = "OnlyLocal"
> > + // AnnotationValueExternalTrafficGlobal Value of annotation to
> specify global (legacy) behaviour > + AnnotationValueExternalTrafficGlobal
> = "Global" > + // AnnotationHealthCheckNodePort Annotation specifying the
> healthcheck nodePort for the service > + // If not specified, annotation is
> created by the service api backend with the allocated nodePort > + // Will
> use user-specified nodePort value if specified by the client > +
> AnnotationHealthCheckNodePort = "service.alpha.kubernetes.io/
> healthcheck-nodeport" Does this let a client request a new NodePort? If so
> this needs to be quota protected (like nodePorts) because they are a finite
> resource and this can allow exhaustion attacks. — You are receiving this
> because you were mentioned. Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29409/files/ b82c028
> <https://github.com/kubernetes/kubernetes/commit/b82c028f77d85ac0be39b510ccdbc2783e43fdc7>
> <https://github.com/kubernetes/kubernetes/commit/
> b82c028f77d85ac0be39b510ccdbc2783e43fdc7>#r77550810>, or mute the thread
> < https://github.com/notifications/unsubscribe- auth/AFVgVLYbJbaO9EXrh0N_
> L1yuemTvxBblks5qnGZwgaJpZM4JSUXH> . — You are receiving this because you
> commented. Reply to this email directly, view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/29409/files/ b82c028
> <https://github.com/kubernetes/kubernetes/commit/b82c028f77d85ac0be39b510ccdbc2783e43fdc7>#r77559231>,
> or mute the thread < https://github.com/notifications/unsubscribe-
> auth/ABG_ p84ktIbrRCwieG7Vn1JDqHLYMvL1ks5qnKSogaJpZM4JSUXH> . — You are
> receiving this because you were mentioned. Reply to this email directly,
> view it on GitHub <https://github.com/kubernetes/kubernetes/pull/
> 29409/files/b82c028f77d85ac0be39b510ccdbc2783e43fdc7#r77565957>, or mute
> the thread <https://github.com/notifications/unsubscribe-
> auth/AFVgVJn9u6GB9Qa9L_uwX2mVCQx79t7Zks5qnNAhgaJpZM4JSUXH> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29409/files/b82c028f77d85ac0be39b510ccdbc2783e43fdc7#r77570919>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxK7hL5xcBunqoJjCLMypMDgRFU_ks5qnOX6gaJpZM4JSUXH>
> .
>